4844: blob encoding version 2

angel-ding-cb opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

This is the implementation based on the 2nd version of [blob encoding spec](https://github.com/ethereum-optimism/optimism/pull/8745/files?short_path=882757e#diff-882757e7a296e749f616fb3e5972b2f8041dc9c515876529981fddd38e73a6a7). 

**Tests**

Tested small and large data inputs, as well as edge cases.


**Metadata**

- Specs #[[Link to Spec](https://github.com/ethereum-optimism/optimism/pull/8745/files?short_path=882757e#diff-882757e7a296e749f616fb3e5972b2f8041dc9c515876529981fddd38e73a6a7)]

angel-ding-cb committed:
blob encoding version 2

tynes commented on op-service/eth/blob_test.go:
```diff
@@ -32,10 +33,63 @@ func TestBlobEncodeDecode(t *testing.T) {
+	// the first field element is filled and no data remains
+	data := Data(make([]byte, 128))
+	data[127] = 0xFF
+
```
We usually use `"github.com/stretchr/testify/require"` for test assertions.

You can do `require.NoError(b.FromData(data), "failed to encode bytes")`

roberto-bayardo replied:
updated to use testify/require

tynes replied:
thank you!

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// It reads 31 bytes and then 1 byte for three times and then 31 bytes of data from the input.
+// For all the 31 bytes of data, they are encoded into [1:32] bytes of each field element.
+// For the extra 3 bytes read from the input, they are encoded into the 1 bytes of data on the top of each field element.
+// So for each field (4 field elements), 3 bytes are encoded into 4 bytes where the highest order bit is set to 0
```
```suggestion
//
// In each round we perform 7 reads of input of lengths (31,1,31,1,31,1,31) bytes respectively for a total of 127
// bytes. This data is encoded into the next 4 field elements of the output by placing each of the 4x31
// byte chunks into bytes [1:32] of its respective field element. The three single byte chunks (24 bits) are
// split into 4x6-bit chunks, each of which is written into the top most byte of its respective field element,
// leaving the top 2 bits of each field element empty to avoid modulus overflow.
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+
+	// encode the length as big-endian uint24 into [2:5] bytes of the first field element
+	if len(data) < 1<<24 {
+		// Zero out any trailing data in the buffer if any
+		b[2] = byte((len(data) >> 16) & 0xFF) // Most significant byte
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
```
Why not use what was there before? It's more concise & probably more efficient.
```suggestion
	binary.LittleEndian.PutUint32(b[1:5], uint32(len(data)))
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	b[1] = EncodingVersion
+
+	// encode the length as big-endian uint24 into [2:5] bytes of the first field element
+	if len(data) < 1<<24 {
+		// Zero out any trailing data in the buffer if any
+		b[2] = byte((len(data) >> 16) & 0xFF) // Most significant byte
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
+	} else {
+		return fmt.Errorf("Error: length_rollup_data is too large")
+	}
```
It's more idiomatic in go to nest only the exceptional case, e.g. (with some other edits/simplifications for clarity): 
```suggestion
    ilen := uint32(len(data))
    if (ilen >> 24) != 0 {
        return fmt.Errorf("unexpected bits in top byte of length field. len=%v", ilen)
    }
    b[2] = byte(ilen >> 16)
    b[3] = byte(ilen >> 8)
    b[4] = byte(ilen)
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+		return nil
+	}
+
+	for fieldNumber := 1; fieldNumber < 1024; fieldNumber++ {
```
```suggestion
	for round := 1; round < 1024; round++ {
```
You might also define a "NumRounds" constant with value 1024 and use this here and in the decoder too.

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
 
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	if offset >= len(data) {
+		// If the offset is at or beyond the end of data, return a new byte array of numBytes length
```
cute, I was wondering how you were dealing with short inputs... I would put this detail about how short reads in the godoc above the function as it's important to understanding how the encoding works. 

Something like "read(..) always returns an array of length exactly numBytes, appropriately padded with zeros should there be not enough input"

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	offset := 0
+	var buffer []byte
+
+	// encode the first 27 + 1 bytes of data into remaining bytes of first field element
```
```suggestion
	// round 0 is special cased here since we can only encode the first (31-5)+1=27 bytes of input into
	//  the first field element due to version/length encoding already occupying the first 5 bytes.
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+
+	// check the version
+	if firstField[1] != EncodingVersion {
+		return nil, fmt.Errorf("invalid blob, expected version %d, got %d", EncodingVersion, firstField[0])
```
```suggestion
		return nil, fmt.Errorf("invalid blob, expected version %d, got %d", EncodingVersion, firstField[1])
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
-	if dataLen > uint32(len(data)) {
+
+	// decode the 3-byte length prefix into 4-byte integer
+	var dataLen int32
```
```suggestion
	var dataLen uint32
```

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	var dataLen int32
+
+	// Assuming b[2], b[3], and b[4] contain the encoded length in big-endian format
+	dataLen = int32(b[2]) << 16 // Shift the most significant byte 16 bits to the left
```
int32 -> uint32 here & below

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	b.write(buffer, 5)
+	pointer := 32 // manually set the pointer to the next field element
+
+	clearBuffer(buffer)
```
I don't think you need these clearBuffer calls because read() always returns a new array.

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -131,3 +272,9 @@ func (b *Blob) Clear() {
 	}
 }
+
+func clearBuffer(buffer []byte) {
```
you can get rid of this once you get rid of the clearBuffer calls as suggested above.

roberto-bayardo commented on op-service/eth/blob.go:
```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	}
+
+	// Create a new byte array and copy the data from the original slice
+	byteArray := make([]byte, actualNumBytes)
```
Preserve the invariant that the returned byte array is always exactly "numBytes".  While it works either way, I think this makes it easier to reason about.  You could even move the byteArray creation to the very top of the function to make this exceptionally clear, and return it even in the offest >= len(data) case.

```suggestion
	byteArray := make([]byte, numBytes)
```

Commit:
- cosmetic cleanup
- use testify/require for testing
- improve big blob test case
- some performance improvements

tynes commented on op-service/eth/blob_test.go:
```diff
@@ -54,16 +94,20 @@ func TestInvalidBlobDecoding(t *testing.T) {
-	}
+	b[VersionOffset] = 0x01 // invalid encoding version
+	_, err = b.ToData()
+	require.Error(t, err)
```
Can we use the following method to assert specifically on certain errors? https://pkg.go.dev/github.com/stretchr/testify/require#ErrorIs

If we break out the errors into their own types then this becomes much easier. We should have test coverage of each error case

roberto-bayardo replied:
done

tynes commented on op-service/eth/blob.go:
```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+func (b *Blob) decodeFieldElement(opos, ipos int, encodedByte []byte, output []byte) (int, int, error) {
+	// two highest order bits of the first byte of each field element should always be 0
+	if b[ipos]&(0xC0) != 0 {
+		return 0, 0, fmt.Errorf("invalid blob, field element %d has a high order bit set", ipos)
```
I do not see any test coverage of this error

roberto-bayardo replied:
see first case of TestInvalidBlobDecoding

tynes commented on op-service/eth/blob_test.go:
```diff
@@ -1,7 +1,10 @@
+
+	"github.com/stretchr/testify/require"
 )
 
```
It would be nice to have some fuzz roundtrip tests for encode/decode

roberto-bayardo replied:
added

tynes commented on op-service/eth/blob.go:
```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+// For only the very first output field, bytes [1:5] are used to encode the version and the length
+// of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
```
There are 2 checks for the size, are both necessary?

tynes replied:
Both error messages are the same so its not possible to distinguish a difference between the 2 cases. If both are required, we should modify the error messages so they are distinct and then test for both cases

roberto-bayardo replied:
the second isn't necessary, was more for documentation that the high order byte will never be set and we can always squeeze this into 3 bytes. removed & replaced with a comment.

tynes commented on op-service/eth/blob.go:
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+// reassembleBytes takes the 4x6-bit chunks from encodedByte, reassembles them into 3 bytes of
+// output, and places them in their appropriate output positions.
+func reassembleBytes(opos int, encodedByte []byte, output []byte) int {
+	opos-- // account for fact that we don't output a 128th byte
```
Should we be worried about underflow here?

roberto-bayardo replied:
reassembleBytes is only called after processing 4 field elements which advances opos by 128.  Max lookback here is within that window, so unless someone introduces a coding error or starts calling this from elsewhere, underflow can't happen.  Could extend this func to also return error if you want me to add an explicit test just in case.

tynes commented on op-service/eth/blob_test.go:
```diff
@@ -72,7 +116,18 @@ func TestTooLongDataEncoding(t *testing.T) {
+	require.Error(t, err)
+}
+
+func FuzzEncodeDecodeBlob(f *testing.F) {
```
You will need to add a call to run this fuzz test in the makefile under `fuzz`

https://github.com/ethereum-optimism/optimism/blob/develop/op-service/Makefile

roberto-bayardo replied:
done

tynes commented on op-service/eth/blob.go:
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+// of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
 		return fmt.Errorf("data is too large for blob. len=%v", len(data))
```
I am sorry if this is annoying but I don't see any test coverage of this error case

roberto-bayardo replied:
no worries :) There is coverage though, see TestTooLongDataEncoding test.   I just updated it to return a specific error type to test for.

tynes commented on op-service/eth/blob.go:
```diff
@@ -70,60 +79,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
-		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-offset)
+
+	if ipos < len(data) {
+		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-ipos)
```
Is there test coverage of this error case?

roberto-bayardo replied:
no because it's one of those errors that shouldn't ever trigger, the topmost check should catch all length issues. this is more for catching coding errors if the code is modified.

tynes commented on op-service/eth/blob.go:
```diff
@@ -14,7 +14,16 @@ import (
+)
+
+var (
+	InvalidFieldElementError    = errors.New("invalid field element")
```
nit: errors are usually prefixed with `Err` and we may want to prefix with `ErrBlob` because this is in the `eth` package

roberto-bayardo replied:
added ErrBlob prefix

roberto-bayardo commented on op-service/Makefile:
```diff
@@ -14,4 +14,6 @@ fuzz:
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadUnmarshal ./eth
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV1 ./eth
 	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV2 ./eth
+	go test -run NOTAREALTEST -v -fuzztime 10s -fuzz FuzzExecutionPayloadMarshalUnmarshalV2 ./eth
```
@tynes is this duplicate test here intentional or should I fix?

tynes replied:
The rabbit is correct, do you mind removing it?

roberto-bayardo replied:
done

trianglesphere reviewed (commented):
The parsing looks pretty good to me & I like the round trip tests. One class of tests I'd like to see is a one way test or test vectors. i.e. Given a known input data, we should expect it to yield a known output. We should check both ways (i.e. given some data, expect a blob output & given a blob, expect some data), but it's nice to have the specific output which can manually be verified.

roberto-bayardo commented:
> One class of tests I'd like to see is a one way test or test vectors. i.e. Given a known input data, we should expect it to yield a known output. We should check both ways (i.e. given some data, expect a blob output & given a blob, expect some data), but it's nice to have the specific output which can manually be verified.

OK to add test vectors as a followup PR? I created an issue for it to make sure it doesn't get dropped and added a TODO referencing it in blob_test.go.

trianglesphere reviewed (approved):
Doing 1 way tests is ok as a follow up

protolambda commented on op-service/eth/blob.go:
```diff
@@ -70,60 +80,193 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+// exactly numBytes, even if there wasn't enough data to read. In the short-read case, any trailing
+// bytes are zero padded.
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	byteArray := make([]byte, numBytes)
+	if offset >= len(data) {
+		return byteArray, len(data)
+	}
```
We should not be allocating a new slice of bytes for every read call.
If we make `read` internal` to the decoding function,  then we can safely avoid the `ipos` input/output, and just return a slice of the original  bytes. This then makes it not allocate so much, and much less verbose. I'll submit a quick PR.

protolambda replied:
https://github.com/roberto-bayardo/optimism/pull/10

protolambda committed:
op-service: blob encoding nits (#10)
